app/org/sagebionetworks/bridge/services/HealthDataService.java

```diff
@@ -27,22 +51,200 @@
+        this.strictValidationHandler = strictValidationHandler;
+    }
+
+    /**
+     * Transcribe Consent Handler, which should be called Transcribe Participant Options Handler. This transcribes
+     * external ID, data groups, and sharing scope into the health data record.
+     */
+    @Autowired
+    public final void setTranscribeConsentHandler(TranscribeConsentHandler transcribeConsentHandler) {
+        this.transcribeConsentHandler = transcribeConsentHandler;
+    }
+
+    /** Upload Artifacts Handler, which finalizes the record and attachments and uploads them to DDB and S3. */
+    @Autowired
+    public final void setUploadArtifactsHandler(UploadArtifactsHandler uploadArtifactsHandler) {
+        this.uploadArtifactsHandler = uploadArtifactsHandler;
+    }
+
+    /* HEALTH DATA SUBMISSION */
+
+    /**
+     * Synchronous health data API. Used to submit small health data payloads (such as survey responses) without
+     * incurring the overhead of creating a bunch of small files to upload to S3.
+     */
+    public HealthDataRecord submitHealthData(StudyIdentifier studyId, StudyParticipant participant,
+            HealthDataSubmission healthDataSubmission) throws JsonProcessingException {
+        // validate health data submission
+        if (healthDataSubmission == null) {
+            throw new InvalidEntityException("Health data submission cannot be null");
+        }
+        Validate.entityThrowingException(HealthDataSubmissionValidator.INSTANCE, healthDataSubmission);
+
+        // sanitize field names in the data node
+        JsonNode sanitizedData = sanitizeFieldNames(healthDataSubmission.getData());
+
+        // Filter data fields and attachments based on schema fields.
+        UploadSchema schema = schemaService.getUploadSchemaByIdAndRev(studyId, healthDataSubmission.getSchemaId(),
+                healthDataSubmission.getSchemaRevision());
+        ObjectNode filteredData = BridgeObjectMapper.get().createObjectNode();
+        Map<String, byte[]> attachmentMap = new HashMap<>();
+        filterAttachments(schema, sanitizedData, filteredData, attachmentMap);
+
+        // construct health data record
+        HealthDataRecord record = makeRecordFromSubmission(studyId, participant, healthDataSubmission, filteredData);
+
+        // Construct UploadValidationContext for the remaining upload handlers. We don't need all the fields, just the
+        // ones that these handlers will be using.
+        UploadValidationContext uploadValidationContext = new UploadValidationContext();
+        uploadValidationContext.setAttachmentsByFieldName(attachmentMap);
+        uploadValidationContext.setHealthCode(participant.getHealthCode());
+        uploadValidationContext.setHealthDataRecord(record);
+        uploadValidationContext.setStudy(studyId);
+
+        // Strict Validation Handler. If this throws, this is an invalid upload (400).
+        try {
+            strictValidationHandler.handle(uploadValidationContext);
```

Comment:
The sequence here of strictValidationHandler -> transcribeConsentHandler -> uploadArtifactsHandler is the same as the end of uploadValidationHandlerList. Would it make sense to create something like a healthDataValidationHandlerList to capture and reuse this handler sequence?

One challenger here is in this case, we are wanting to short circuit if strictValidationHandler fails

Reply:
Personally, I think the configured handler list, while clever, was actually over-engineered. It would be a lot more useful if we had multiple upload paths and wanted to be able to hot swap handlers quickly. As it is, the 8 handlers have only changed once, and it was to create a new version of the IosSchemaValidationHandler. The Synchronous Upload API only has 3 steps (4 if we want to encode all the prep work as a handler), so I don't think it needs something as sophisticated as UploadValidationTask.

If we do need something more sophisticated, it wouldn't be difficult to refactor this code there and then.